### PR TITLE
[2839] - live location lookup

### PR DESCRIPTION
**Pull request by TobyRet**
### Context
When a user searches by 'postcode, town or city' on the location filter page, location suggestions are now displayed.

location

### Guidance to review
I expect _DependaSau_ will break this

### CODE REVIEW

File: app/models/location_suggestion.rb [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+
+  base_uri "https://maps.googleapis.com"
+
+  class << self
+    def suggest(input)
+      query = build_query(input)
+
+      response = get("/maps/api/place/autocomplete/json?#{query.to_query}")
+
+      if response.success?
```
Comment (TobyRet) [inline app/models/location_suggestion.rb]: More / better error handling required?

**Comment by TobyRet** [comment]
[[M] location filter: Live address lookup](https://trello.com/c/5j7Vwnpw/2839-m-location-filter-live-address-lookup)

File: app/services/location_suggestion.rb [hunk 2]
```diff
@@ -0,0 +1,29 @@
+class LocationSuggestion
+  include HTTParty
+
+  base_uri "https://maps.googleapis.com"
```
Comment (defong) [inline app/services/location_suggestion.rb]: Maybe stick these into settings?

File: app/services/location_suggestion.rb [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+
+  class << self
+    def suggest(input)
+      query = build_query(input)
+
+      response = get("/maps/api/place/autocomplete/json?#{query.to_query}")
```
Comment (defong) [inline app/services/location_suggestion.rb]: Settings as well maybe?

File: app/webpacker/scripts/autocomplete.js [hunk 4]
```diff
@@ -32,7 +32,7 @@ const initAutocomplete = ($el, $input) => {
   const el = document.getElementById($el);
   const inputValueTemplate = result => (typeof result === "string" ? result : result && result.name);
   const suggestionTemplate = result =>
-    typeof result === "string" ? result : result && `${result.name} (${result.providerCode})`;
+    typeof result === "string" ? result : result && `${result.name} (${result.code})`;
```
Comment (defong) [inline app/webpacker/scripts/autocomplete.js]: Does this change break the provide code?

File: app/views/result_filters/location/new.html.erb [hunk 5]
```diff
@@ -50,12 +50,14 @@
                 <%=
                   form.text_field(
                     :lq,
+                    id: 'location',
                     value: "#{request.params[:lq]}",
                     class: "govuk-input",
-                    data: {qa: "location-query"},
+                    data: {qa: "location-query", url: "/location-suggestions"},
```
Comment (defong) [inline app/views/result_filters/location/new.html.erb]: Is this url same as location_suggestions_path (railsy one)

File: app/webpacker/scripts/autocomplete.js [hunk 6]
```diff
@@ -32,7 +32,7 @@ const initAutocomplete = ($el, $input) => {
   const el = document.getElementById($el);
   const inputValueTemplate = result => (typeof result === "string" ? result : result && result.name);
   const suggestionTemplate = result =>
```
Comment (defong) [inline app/webpacker/scripts/autocomplete.js]: inputValueTemplate and suggestionTemplate functions can be move out of initAutocomplete.

File: app/views/result_filters/location/new.html.erb [hunk 7]
```diff
@@ -38,7 +38,7 @@
               <%= form.label :l_1, "By postcode, town or city", class: "govuk-label govuk-radios__label" %>
             </div>
             <div
-              class="govuk-radios__conditional <%= "govuk-radios__conditional--hidden" unless location_error? %>"
+              class="govuk-radios__conditional <%= "govuk-radios__conditional--hidden" unless request.params[:lq] || location_error? %>"
```
Comment (aliuk2012) [inline app/views/result_filters/location/new.html.erb]: Do we even need to include `location_error?`? shouldn't we always show the conditional fields if the parent radio button has been set.

Comment (aliuk2012) [reply app/views/result_filters/location/new.html.erb]: Is this the same change as whats coming in #92 ?

Comment (aliuk2012) [inline app/views/result_filters/location/new.html.erb]:
```suggestion
              class="govuk-radios__conditional <%= "govuk-radios__conditional--hidden" unless request.params[:lq] %>"
```

File: app/services/location_suggestion.rb (hunk 2 above)
Comment (TobyRet) [reply app/services/location_suggestion.rb]: Woops. Will do

File: app/webpacker/scripts/autocomplete.js (hunk 4 above)
Comment (TobyRet) [reply app/webpacker/scripts/autocomplete.js]: It does the opposite. It unbreaks it 😄 ... The code highlighted in red is C# . I'll add as a separate commit

File: app/views/result_filters/location/new.html.erb (hunk 5 above)
Comment (TobyRet) [reply app/views/result_filters/location/new.html.erb]: Yep. You rails ninja

File: app/webpacker/scripts/autocomplete.js (hunk 6 above)
Comment (TobyRet) [reply app/webpacker/scripts/autocomplete.js]: I don't see why you would want to do this? They should be scoped to the function as they are not used elsewhere

File: app/views/result_filters/location/new.html.erb (hunk 7 above)
Comment (TobyRet) [reply app/views/result_filters/location/new.html.erb]: Ah yes true

File: app/webpacker/scripts/autocomplete.js (hunk 6 above)
Comment (aliuk2012) [reply app/webpacker/scripts/autocomplete.js]: I agree with @TobyRet, if its not being used else where then it should be scoped to the function.

File: app/views/result_filters/location/new.html.erb [hunk 8]
```diff
@@ -121,7 +123,7 @@
                                     id: "provider",
                                     value: params[:query],
                                     class: "govuk-input",
-                                    data: {qa: "provider-search"} %>
+                                    data: {qa: "provider-search", url: "/provider-suggestions"} %>
```
Comment (aliuk2012) [inline app/views/result_filters/location/new.html.erb]: Another rails path?

File: app/models/location_suggestion.rb (hunk 1 above)
Comment (TobyRet) [reply app/models/location_suggestion.rb]: Any other exceptions will automatically go to Sentry ...

File: app/views/result_filters/location/new.html.erb (hunk 8 above)
Comment (TobyRet) [reply app/views/result_filters/location/new.html.erb]: Yep got that

File: app/webpacker/scripts/autocomplete.js [hunk 9]
```diff
@@ -43,7 +43,7 @@ const initAutocomplete = ($el, $input) => {
         name: input.name,
         defaultValue: input.value,
         minLength: 3,
-        source: request("/provider-suggestions"),
+        source: request(input.getAttribute("data-url")),
```
Comment (aliuk2012) [inline app/webpacker/scripts/autocomplete.js]: could this be open to xss? Just thinking if the attribute is replaced with something malicious then the lookup could be use as a keylogger.

Comment (TobyRet) [reply app/webpacker/scripts/autocomplete.js]: Path is now being passed as param to initAutocomplete function

**Pull request merged**